src/main/java/net/minecraftforge/client/GeneratorTypeManager.java

```diff
@@ -0,0 +1,196 @@
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Register a GeneratorType and associated EditScreenFactory with the GeneratorTypeManager
+     *
+     * @param generatorType The GeneratorType to be added
+     * @param editScreen    The EditScreenFactory to be associated with given GeneratorType
+     * @return true if the GeneratorType and EditScreenFactory was successfully added
+     */
+    public synchronized boolean register(BiomeGeneratorTypeScreens generatorType, BiomeGeneratorTypeScreens.IFactory editScreen) {
+        if (register(generatorType)) {
+            editScreens.put(generatorType, editScreen);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Set the EditScreenFactory for a given GeneratorType, overriding any previously set factory
+     *
+     * @param generatorType The GeneratorType to associate with the given EditScreenFactory
+     * @param editScreen    The EditScreenFactory to associate with the given GeneratorType
+     * @return true if the EditScreenFactory was successfully set for the GeneratorType
+     */
+    public synchronized boolean setEditScreen(BiomeGeneratorTypeScreens generatorType, BiomeGeneratorTypeScreens.IFactory editScreen) {
+        if (generatorTypes.contains(generatorType)) {
+            editScreens.put(generatorType, editScreen);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Set the default GeneratorType
+     *
+     * @param generatorType The GeneratorType to use as the default
+     */
+    public synchronized void setDefaultGeneratorType(@Nonnull BiomeGeneratorTypeScreens generatorType) {
+        this.defaultGeneratorType = generatorType;
+    }
+
+    /**
+     * Set the index of the given GeneratorType in the list of GeneratorType
+     * No guarantees can be made about the resulting order of the GeneratorTypes list when multiple mods
+     * set the index of a GeneratorType
+     *
+     * @param generatorType The GeneratorType to set the index of
+     * @param index         The index to set for the given GeneratorType
+     * @return true if the index was successfully set or the GeneratorType was already placed at the given index
+     */
+    public synchronized boolean setIndex(BiomeGeneratorTypeScreens generatorType, int index) {
+        int currentIndex = generatorTypes.indexOf(generatorType);
+        if (currentIndex == -1) {
+            return false;
+        }
+        if (currentIndex == index) {
+            return true;
+        }
+        generatorTypes.remove(currentIndex);
+        generatorTypes.add(index, generatorType);
+        return true;
+    }
+
+    public static GeneratorTypeManager get() {
+        return INSTANCE;
+    }
+
+    // Used to access the list/maps of GeneratorTypes in BiomeGeneratorTypeScreens so that they can be
+    // copied to the manager.
+    private static class VanillaGeneratorTypeAccess extends BiomeGeneratorTypeScreens {
```

Comment:
A possible simplification: this class exists only to access two fields, `field_239068_c_` and `field_239069_d_`, both of which could just be AT'd to public instead. This feels like a very roundabout way to do that, using the quirk of a protected static field. It's also named as if it was a mixin `@Accessor` which confused me at first. Might be simpler just to AT both fields as there's not really any potential implementation hiding since any mod subclassing `BiomeGeneratorTypeScreens` will be able to access them anyway.

Reply:
True. In my personal projects as a rule of thumb if I can achieve something without AT's I go with that even if it's a bit jank-looking. I can see how simple == maintainable might be preferable in this project.

Reply:
Honestly, if anything, I feel like these fields should be made private and their contents copied to the GeneratorTypeManager in a static block or something. This'd make the patch larger and possibly be considered breaking.

Making something public that no-one else is supposed to use doesn't _feel_ like the right move either. Not sure what the correct way forward is :/ More guidance would be appreciated

Reply:
Marking as resolved as this now outdated, however, I've stuck with the pattern in the revisions (see GeneratorOptionList) so feel free to raise again for discussion/guidance